docker/importer/importer.py

```diff
@@ -307,50 +315,82 @@ def _process_updates_git(self, source_repo):
     source_repo.last_synced_hash = str(repo.head.target)
     source_repo.put()
 
-  def _process_updates_bucket(self, source_repo):
+    logging.info("Finish processing git: %s", source_repo.name)
+
+  def _process_updates_bucket(self, source_repo: osv.SourceRepository):
     """Process updates from bucket."""
     # TODO(ochang): Use Pub/Sub change notifications for more efficient
     # processing.
+    logging.info("Begin processing bucket: %s", source_repo.name)
 
     ignore_last_import_time = source_repo.ignore_last_import_time
     if ignore_last_import_time:
       source_repo.ignore_last_import_time = False
       source_repo.put()
 
+    # First retrieve a list of files to parallel download
     storage_client = storage.Client()
-    for blob in storage_client.list_blobs(source_repo.bucket):
-      if not _is_vulnerability_file(source_repo, blob.name):
-        continue
-
-      logging.info('Bucket entry triggered for for %s/%s', source_repo.bucket,
-                   blob.name)
-      blob_bytes = blob.download_as_bytes()
-      try:
-        vulnerabilities = osv.parse_vulnerabilities_from_data(
-            blob_bytes,
-            os.path.splitext(blob.name)[1])
-      except Exception as e:
-        logging.error('Failed to parse vulnerability %s: %s', blob.name, e)
-        continue
-
-      def unchanged_vuln_exist(vuln):
-        bug = osv.Bug.get_by_id(vuln.id)
-        # Both times are not timezone aware
-        return bug and bug.import_last_modified == vuln.modified.ToDatetime()
-
-      if not ignore_last_import_time and all(
-          unchanged_vuln_exist(vuln) for vuln in vulnerabilities):
-        logging.info('Skipping updates for %s as modified date unchanged.',
-                     blob.name)
-        continue
-
-      original_sha256 = osv.sha256_bytes(blob_bytes)
-      self._request_analysis_external(source_repo, original_sha256, blob.name)
+    listed_blob_names = [
+        blob.name for blob in (storage_client.list_blobs(source_repo.bucket))
+    ]
+    batched_blob_names = [
+        listed_blob_names[i:i + _BUCKET_BATCH_SIZE]
+        for i in range(0, len(listed_blob_names), _BUCKET_BATCH_SIZE)
+    ]
+
+    # Setup storage client
+    def thread_init(func):
+      func.client = storage.Client()
+
+    with multiprocessing.Pool(
+        _BUCKET_THREAD_POOL,
+        initializer=thread_init,
+        initargs=(convert_blob_to_vuln,)) as processing_pool:
+      for batch_blob in batched_blob_names:
+        conv_func_partial = functools.partial(convert_blob_to_vuln, source_repo)
+        converted_vulns = processing_pool.map(conv_func_partial, batch_blob)
+        # Flatten list[list[]] to list[]
+        vulnerabilities = [
+            inner for outer in converted_vulns for inner in outer
+        ]
+
+        # Batch query the bugs
+        vuln_ids = [x.id for x, _, _ in vulnerabilities]
+        existing_vulns = osv.Bug.query(osv.Bug.db_id.IN(vuln_ids))
```

Comment:
What about ndb initialization? According to this: https://googleapis.dev/python/python-ndb/latest/migrating.html#setting-up-a-connection

it's not thread safe.

Reply:
This is still ran on the main thread I believe. The only part that is multithreaded is downloading blobs from the google cloud bucket. Each client there is initialized per thread.